src/storage/src/upsert/types.rs

```diff
@@ -703,48 +998,56 @@ where
     ///    merge operands are merged by the backend.
     async fn multi_merge<P>(&mut self, merges: P) -> Result<MergeStats, anyhow::Error>
     where
-        P: IntoIterator<Item = (UpsertKey, MergeValue<StateValue<O>>)>;
+        P: IntoIterator<Item = (UpsertKey, MergeValue<StateValue<T, O>>)>;
 }
 
-/// A function that merges a set of updates for a key into the existing value for the key, expected
-/// to only be used during the snapshotting-phase of an upsert operator. This is called by the
-/// backend implementation when it has accumulated a set of updates for a key, and needs to merge
-/// them into the existing value for the key.
+/// A function that merges a set of updates for a key into the existing value
+/// for the key. This is called by the backend implementation when it has
+/// accumulated a set of updates for a key, and needs to merge them into the
+/// existing value for the key.
 ///
 /// The function is called with the following arguments:
 /// - The key for which the merge is being performed.
 /// - An iterator over any current value and merge operands queued for the key.
+///
 /// The function should return the new value for the key after merging all the updates.
-pub(crate) fn snapshot_merge_function<O>(
+pub(crate) fn consolidating_merge_function<T, O>(
     _key: UpsertKey,
-    updates: impl Iterator<Item = StateValue<O>>,
-) -> StateValue<O>
+    updates: impl Iterator<Item = StateValue<T, O>>,
+) -> StateValue<T, O>
 where
     O: Default,
+    T: std::cmp::Eq,
 {
-    let mut current = Default::default();
-    assert!(
-        matches!(current, StateValue::Snapshotting(_)),
-        "merge_function called with non-snapshotting default"
-    );
+    let mut current: StateValue<T, O> = Default::default();
+
     for update in updates {
-        assert!(
-            matches!(update, StateValue::Snapshotting(_)),
-            "merge_function called with non-snapshot update"
-        );
-        current.merge_update_state(&update);
+        match update {
+            StateValue::Consolidating(_) => {
+                current.merge_update_state(&update);
+            }
+            StateValue::Value(_) => {
+                // This branch is more expensive, but we hopefully rarely hit
+                // it.
+                if let Some((finalized_value, _order)) = update.into_finalized_value() {
+                    let mut update = StateValue::default();
+                    update.merge_update(finalized_value, 1, upsert_bincode_opts(), &mut Vec::new());
```

Comment:
Can you lift the auxiliary vector out of the loop so that we reuse it?

Reply:
will do